npe2/_plugin_manager.py

```diff
@@ -52,125 +53,116 @@ def __getitem__(self, index: Union[int, slice]) -> Set[Interval[T]]:
-        self._readers.clear()
-        self._writers_by_type.clear()
-        self._writers_by_command.clear()
-
-        for result in PluginManifest.discover():
+
+        for result in PluginManifest.discover(paths=paths):
             if result.manifest is None:
                 continue
             mf = result.manifest
-            if filter_by_key and (mf.key not in filter_by_key):
-                continue
             self._manifests[mf.key] = mf
             if mf.contributions:
-                for cmd in mf.contributions.commands or []:
-                    self._commands[cmd.id] = (cmd, mf.key)
-                for subm in mf.contributions.submenus or []:
-                    self._submenus[subm.id] = subm
-                for theme in mf.contributions.themes or []:
-                    self._themes[theme.id] = theme
-                for reader in mf.contributions.readers or []:
-                    for pattern in reader.filename_patterns:
-                        self._readers[pattern].append(reader)
-                    if reader.accepts_directories:
-                        self._readers[""].append(reader)
-                for writer in mf.contributions.writers or []:
-                    self._writers_by_command[writer.command].append(writer)
-                    for c in writer.layer_type_constraints():
-                        self._writers_by_type[c.layer_type].addi(*c.bounds, writer)
+                self._contrib.index_contributions(mf.contributions, mf.key)
 
     def iter_menu(self, menu_key: str) -> Iterator[MenuItem]:
         for mf in self._manifests.values():
             if mf.contributions:
                 yield from getattr(mf.contributions.menus, menu_key, [])
 
-    def get_command(self, command_id: str) -> CommandContribution:
-        return self._commands[command_id][0]
-
-    def get_manifest(self, command_id: str) -> PluginManifest:
-        mf = self._commands[command_id][1]
-        return self._manifests[mf]
-
-    def get_submenu(self, submenu_id: str) -> SubmenuContribution:
-        return self._submenus[submenu_id]
-
     def activate(self, key: PluginKey) -> PluginContext:
         # TODO: this is an important function... should be carefully considered
         try:
-            pm = self._manifests[key]
+            mf = self._manifests[key]
         except KeyError:
             raise KeyError(f"Cannot activate unrecognized plugin key {key!r}")
 
-        # TODO: prevent "reactivation"
-
         # create the context that will be with this plugin for its lifetime.
-        ctx = PluginContext.get_or_create(key)
+        ctx = self.get_context(key)
+        if ctx._activated:
+            # prevent "reactivation"
+            return ctx
+
         try:
             modules_pre = set(sys.modules)
-            pm.activate(ctx)
+            mf.activate(ctx)
             # store the modules imported by plugin activation
             # (not sure if useful yet... but could be?)
```

Comment:
just looking at some of these comments. Maybe we should get rid of ctx._imports etc if we're not using it.

Reply:
sure, will do